crates/libm/src/math/sincosf.rs

```diff
@@ -23,7 +23,9 @@ const S2PIO2: f32 = 2.0 * PI_2; /* 0x400921FB, 0x54442D18 */
 const S3PIO2: f32 = 3.0 * PI_2; /* 0x4012D97C, 0x7F3321D2 */
 const S4PIO2: f32 = 4.0 * PI_2; /* 0x401921FB, 0x54442D18 */
 
-pub fn sincosf(x: f32) -> (f32, f32) {
+#[inline]
+#[cfg_attr(all(test, assert_no_panic), no_panic::no_panic)]
+pub extern "C" fn sincosf(x: f32) -> (f32, f32) {
```

Comment:
Here should be wrapper for compatibility with C. Maybe something like this:
```
#[inline]
pub extern "C" fn sincosf(x: f32, sin: &mut f32, cos: &mut f32) {
     let (s, c) = sin_cosf(x);
     *sin = s;
     *cos = c;
}

#[inline]
pub fn sin_cosf(x: f32) -> (f32, f32) {
 ...
}
```

Reply:
The main point of `extern "C"` is making the function `#[no_unwind]`.

Reply:
Why do you need Rust compatible wrappers for these APIs ?

Reply:
Ahhhh! @burrbull sorry! I understood your argument wrong! 

You mean, that the signature of these functions does not match the C one, right ?

If that's the case, yes, I agree that this should be fixed, but I haven't gotten around to it. The rust-analyze script reports that this is the case for the following APIs: 

```
[error]: Function "modf" returns unsupported type -> ( f64 , f64 )
[error]: Function "remquo" returns unsupported type -> ( f64 , i32 )
[error]: Function "frexpf" returns unsupported type -> ( f32 , i32 )
[error]: Function "remquof" returns unsupported type -> ( f32 , i32 )
[error]: Function "lgammaf_r" returns unsupported type -> ( f32 , i32 )
[error]: Function "sincosf" returns unsupported type -> ( f32 , f32 )
[error]: Function "modff" returns unsupported type -> ( f32 , f32 )
[error]: Function "lgamma_r" returns unsupported type -> ( f64 , i32 )
[error]: Function "sincos" returns unsupported type -> ( f64 , f64 )
[error]: Function "frexp" returns unsupported type -> ( f64 , i32 )
[error]: Function "modf" returns unsupported type -> ( f64 , f64 )
[error]: Function "remquo" returns unsupported type -> ( f64 , i32 )
[error]: Function "frexpf" returns unsupported type -> ( f32 , i32 )
[error]: Function "remquof" returns unsupported type -> ( f32 , i32 )
[error]: Function "lgammaf_r" returns unsupported type -> ( f32 , i32 )
[error]: Function "sincosf" returns unsupported type -> ( f32 , f32 )
[error]: Function "modff" returns unsupported type -> ( f32 , f32 )
[error]: Function "lgamma_r" returns unsupported type -> ( f64 , i32 )
[error]: Function "sincos" returns unsupported type -> ( f64 , f64 )
[error]: Function "frexp" returns unsupported type -> ( f64 , i32 )
```

Reply:
I really noob in FFI. What signature will this function have for C user?
```
void sincos(double x, double *sin, double *cos);
```

Reply:
Can the pointers in the C API be NULL? Just something to be aware of before exposing to C a function that UBs on NULL input.

Reply:
http://git.musl-libc.org/cgit/musl/tree/src/math/sincos.c#n16

Reply:
Good, the pointers cannot indeed be NULL, however, given the C example, instead of `&mut f32`, `&Cell<f32>` should be used.

Reply:
A cell is not `f32` pointer sized. Maybe we need `NonNull<T>` here ?

Reply:
The closest equivalent to C's `integer_type *` (a nullable mutable pointer to a potentially aliased integer), is `Option<&Cell<integer_type>>`.
The `Option` can be skipped because C implementation does not check against that,  and given that this function is not thread-safe anyways, the `&Cell` could be replaced by a `&mut`, I guess. One last question, though: is `f32` the portable equivalent of `C`'s `double` ?

Reply:
> The closest equivalent to C's numeric_type * (a nullable mutable pointer to a potentially aliased integer), is Option<&Cell<numeric_type>>.

Using a raw pointer:

```rust
let sin = MaybeUninit::uninit();
let cos = MaybeUninit::uninit();
sincos(x, sin.get_mut_ptr(), cos.get_mut_ptr()); 
```

is ok, because a raw pointer *mut T can point to an invalid T - since the T is uninitialized, the T is, in general, invalid. (*)

Using a `&Cell<T>`  , `&T` requires `T` to be _valid_. Since `Cell<T>` stores a _valid_ `T` behind a `UnsafeCell<T>`, and `UnsafeCell<T>` requires its contents to be valid, one cannot use `&Cell<T>` to initialize a memory location for an invalid `T`.

Using `NonNull<T>` this is not an issue either, since `NonNull<T>` stores a `*mut T`, which can point to an invalid `T`. 

cc @ralfjung - I kind of wish `MaybeUninit` would return `NonNull<T>`, so that we can use `NonNull<T>` in these APIs. 

---


(*) the only `T`s that would be valid here are those for which the "all bits are uninitialized" is a valid bit-pattern - but right now this is not specified for any Ts.

Reply:
> One last question, though: is f32 the portable equivalent of C's double ?

No: `f32` and `f64` are not guaranteed to be layout compatible with C's `float` and `double` types. This should use `libc::c_float` and `lib::c_double`, which are guaranteed to be layout compatible with C's `float` and `double` types. There is no Rust type compatible with C's `long double`, nor can we in general support C's `float_t` and `double_t` types either, due to a lack of a `f80` type.

Reply:
Since this is kind of out-of-scope for this PR, I've opened an issue to discuss this: https://github.com/rust-lang-nursery/libm/issues/199 .